compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/calls/Arguments.kt

```diff
@@ -219,12 +218,29 @@ fun Candidate.resolvePlainArgumentType(
     val session = sink.components.session
     val capturedType = prepareCapturedType(argumentType)
 
-    val argumentTypeForApplicabilityCheck =
+    var argumentTypeForApplicabilityCheck =
         if (useNullableArgumentType)
             capturedType.withNullability(ConeNullability.NULLABLE, session.typeContext)
         else
             capturedType
 
+    // If the argument is of functional type and the expected type is a suspend function type, we need to do "suspend conversion."
+    // TODO: should refer to LanguageVersionSettings.SuspendConversion
+    // TODO: should prefer another candidate without suspend conversion when ambiguous
```

Comment:
@mglukhikh I think we can add `useSuspendConversion` to `Candidate`, somewhat similar to `useSam`, and then filter out if some of candidates have it. If this sounds like a good direction, I can also do this in another follow-up PR, and probably at that time, will consider suspend conversion on arguments and callable reference together.

Reply:
Probably yes, this direction looks good.

Reply:
Implemented at the 2nd commit: https://github.com/JetBrains/kotlin/pull/3710/commits/b17f7f51a59a85ef7fd80aec0541c066857a026c